File: concepts/03_errors_and_resources/04_contracts_assertions_and_failure_signaling/julia/test_02_configuration_and_runtime_boundaries.jl
# polyglot-family: errors_and_resources
# polyglot-concept: contracts_assertions_and_failure_signaling
# polyglot-related: languages/julia/language/07_unicode_exceptions_and_resources/
# polyglot-related+: test_052_parsing_and_tryparse_failure_values.jl
#
# 共同问题：可预期失败何时返回值、何时抛异常；类型边界何时拒绝调用。
# 对照观察：tryparse 返回 nothing，parse 抛异常；multiple dispatch 对无适用签名抛 MethodError。

using Test

accept_integer(value::Int) = value

@testset "失败通道由 API 契约决定" begin
    @test tryparse(Int, "bad") === nothing
    @test_throws ArgumentError parse(Int, "bad")
    @test accept_integer(2) == 2
    @test_throws MethodError accept_integer(2.0)
end

File: concepts/03_errors_and_resources/04_contracts_assertions_and_failure_signaling/rust/test_02_configuration_and_runtime_boundaries.rs
// polyglot-family: errors_and_resources
// polyglot-concept: contracts_assertions_and_failure_signaling
// polyglot-related: languages/rust/tests/standard_library/
// polyglot-related+: 14_formatting_parsing_binary_and_serialization/test_106_from_str_and_numeric_parsing.rs
//
// 共同问题：配置文本何时验证；运行期能力缺失如何与程序错误区分。
// 对照观察：边界解析返回 typed `Result`；可选 OS/capability 用 cfg 或 Option，而非捕获任意 panic。

fn parse_port(text: &str) -> Result<u16, &'static str> {
    let port = text.parse::<u16>().map_err(|_| "port is not u16")?;
    if port == 0 {
        Err("port zero is reserved")
    } else {
        Ok(port)
    }
}

#[test]
fn comparison() {
    assert_eq!(parse_port("8080"), Ok(8080));
    assert_eq!(parse_port("bad"), Err("port is not u16"));
    assert_eq!(parse_port("0"), Err("port zero is reserved"));
    assert!(std::thread::available_parallelism().unwrap().get() >= 1);
}
